src/librustc/infer/error_reporting/mod.rs

```diff
@@ -1155,6 +1158,87 @@ impl<'a, 'tcx> InferCtxt<'a, 'tcx> {
             _ => {}
         }
 
+        struct OpaqueTypesVisitor<'tcx> {
+            types: FxHashMap<&'static str, FxHashSet<Span>>,
```

Comment:
Feels excessive, slow, and confusing to use `FxHashMap<&'static str` when `&'static str` is only ever `opaque type` or `closure` -- use two fields instead for each instead?

Reply:
I thought so too, but I can also see this being extended to other things, like type params and would make `fn report` a bit more involved than it is now.

Reply:
I think we should change things to something less structured (e.g. `HashMap`) when extending it is actually necessary. Also, type params seems like unrelated to opaque types.

Reply:
I'm still not a fan of this `HashMap` approach, but if you want to stick to it, please use an `enum` as the key at least so it's less stringly typed.